Calling SecStaticCodeCheckValidity() takes 16ms on WebProcess launch

pvollan opened the pull request:
#### 00634c1b3ea6169be499097a1e8304afbfd984b0
<pre>
Calling SecStaticCodeCheckValidity() takes 16ms on WebProcess launch
<a href="https://bugs.webkit.org/show_bug.cgi?id=256251">https://bugs.webkit.org/show_bug.cgi?id=256251</a>
rdar://108020395

Reviewed by Geoffrey Garen.

Dispatch work on a secondary thread, so we can do other useful work while this XPC communication is in progress.
Local measurements show that this change reduces the blocking of the main thread from ~15ms to &lt; 1ms.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/263817@main">https://commits.webkit.org/263817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/afa7a867a3e11f28654e23620c5c6a5453ba6037

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5562 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7122 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4977 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7146 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3198 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12119 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5064 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6861 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5514 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4508 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4967 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1378 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9072 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5329 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -477,6 +482,11 @@ static void softlinkDataDetectorsFrameworks()
+
+#if HAVE(VIDEO_RESTRICTED_DECODING) && PLATFORM(MAC)
+    if (codeCheckSemaphore)
+        dispatch_semaphore_wait(codeCheckSemaphore.get(), dispatch_time(DISPATCH_TIME_NOW, 50 * NSEC_PER_MSEC));
```
Don't we need to wait forever here? Otherwise, in outlier cases where the trustd check-in takes more than 50ms, we'll get a process with mystery media failures.

pvollan replied:
> Don't we need to wait forever here? Otherwise, in outlier cases where the trustd check-in takes more than 50ms, we'll get a process with mystery media failures.

Ah, that is a good point! Will fix.

Thanks for reviewing!

geoffreygaren commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -279,17 +279,22 @@ static void softlinkDataDetectorsFrameworks()
-            CFRelease(code);
-        }
+        codeCheckSemaphore = adoptOSObject(dispatch_semaphore_create(0));
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), [codeCheckSemaphore = codeCheckSemaphore] {
```
Probably worth comment here that this dispatch is a perf workaround to skip some main thread only work.

cdumez replied:
Doesn't this cause the main thread to wait on a background thread which presumably has lower priority?

geoffreygaren replied:
Yes, the main thread waits. It's a weird thing to do. The reason we do this is that the underlying framework has some arguably arguably ill-advised logging that it performs if and only if it's called on the main thread. So calling it on a secondary thread is a workaround.

That said, probably DISPATCH_QUEUE_PRIORITY_DEFAULT is a similar priority?

pvollan replied:
> Probably worth comment here that this dispatch is a perf workaround to skip some main thread only work.

Ah, good point! I will add a comment.

pvollan replied:
> Doesn't this cause the main thread to wait on a background thread which presumably has lower priority?

That is a good point! Local measurements show that the background thread has finished its work when the wait starts, so perhaps this is ok in this case? If you'd like, I can adjust the priority.

Thanks for reviewing!

cdumez replied:
I think we should use higher priority of the thread. If the work hadn't finished, we wouldn't want the main thread to hang on a lower priority background thread.

geoffreygaren reviewed (approved):
r=me

pvollan committed:
Calling SecStaticCodeCheckValidity() takes 16ms on WebProcess launch
https://bugs.webkit.org/show_bug.cgi?id=256251
rdar://108020395

Reviewed by Geoffrey Garen.

Dispatch work on a secondary thread, so we can do other useful work while this XPC communication is in progress.
Local measurements show that this change reduces the blocking of the main thread from ~15ms to < 1ms.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: https://commits.webkit.org/263817@main
